### PR TITLE
add environment and x509

### PR DESCRIPTION
**Pull request by julcollas**
Environment module
x509 module

### CODE REVIEW

File: testinfra/modules/environment.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        "EDITOR": "vim",
+        "SHELL": "/bin/bash",
+        [...]
+    }
+    """
+
+    def __call__(self):
+        cmd = "python -c 'import json, os;print(json.dumps(dict(os.environ)))'"
+        return json.loads(self.check_output(cmd))
```
Comment (philpep) [inline testinfra/modules/environment.py]: I'd rather rely on `env` command since python may not be available on target host.

**Comment by varac** [comment]
@julcollas What's the state of this MR ? I'd love to see the x509 module merged!

**Comment by julcollas** [comment]
Almost 1 year but, I made the changes :)

File: testinfra/modules/environment.py (hunk 1 above)
Comment (julcollas) [reply testinfra/modules/environment.py]: changes have been commited to rely on env

**Comment by varac** [comment]
@julcollas Did you notice the failing travis test ?

File: testinfra/modules/x509.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        """Return enddate as datetime of certificate
+
+        >>> host.x509('/etc/pki/tls/certs/dummy.crt').enddate
+        """
+
+        raise NotImplementedError
+
+    @property
+    def serial(self):
+        """Return enddate as datetime of certificate
```
Comment (varac) [inline testinfra/modules/x509.py]: I guess this is a copy&paste error from l. 56 ?

**Comment by varac** [comment]
Great, thx, looks like it's ready to merge !

**Comment by varac** [comment]
What's going on - all checks passed and approved, who's going to merge it ?

File: test/test_modules.py [hunk 3]
```diff
@@ -483,6 +483,22 @@ def test_pip_package(host):
     assert int(outdated['latest'].split('.')[0]) > 2
 
 
+def test_x509(host):
+    host.check_output('apt-get -y update')
+    host.check_output('apt-get -y install curl')
+    host.check_output(
+        'curl -so /tmp/crt https://letsencrypt.org/certs/isrgrootx1.pem.txt'
+    )
```
Comment (philpep) [inline test/test_modules.py]: I'd suggest using local (in docker container) certificates from /etc/ssl instead.

File: test/test_modules.py [hunk 4]
```diff
@@ -483,6 +483,22 @@ def test_pip_package(host):
...
+    assert host.file('/tmp/crt').is_file
+    assert 'ISRG Root' in host.x509('/tmp/crt').subject
+    in_ten_days = datetime.datetime.now() + datetime.timedelta(days=10)
+    assert host.x509('/tmp/crt').enddate > in_ten_days
```
Comment (philpep) [inline test/test_modules.py]: Maybe generate a certificate in dockerfile with a fixed end date for this test ?

File: test/test_modules.py [hunk 5]
```diff
@@ -496,6 +512,7 @@ def test_iptables(host):
     assert ssh_rule_str in input_rules
     assert vip_redirect_rule_str in nat_rules
     assert vip_redirect_rule_str in nat_prerouting_rules
+
```
Comment (philpep) [inline test/test_modules.py]: unrelated hunk

File: testinfra/modules/__init__.py [hunk 6]
```diff
@@ -29,7 +31,6 @@
     'pip_package': 'pip:PipPackage',
     'process': 'process:Process',
     'puppet_resource': 'puppet:PuppetResource',
-    'facter': 'puppet:Facter',
```
Comment (philpep) [inline testinfra/modules/__init__.py]: unrelated hunk

File: testinfra/modules/environment.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        "EDITOR": "vim",
+        "SHELL": "/bin/bash",
+        [...]
+    }
+    """
+
+    def __call__(self):
+        ret_val = dict(
+            i.split('=') for i in self.check_output('env').rstrip().split('\n')
+        )
```
Comment (philpep) [inline testinfra/modules/environment.py]: IIRC, `check_output()` already do the `rstrip()`.
Also this behave unexpectedly when using environment variable with a `=` in it (example: export FOO=BAR=ZAZ)

I'd suggest instead:
```return dict(i.split('=', 1) for i in self.check_output('env').splitlines()```

But there still issues with environment variables containing a '\n', for this I'd suggest using the `-0` option of `env` (replacing newlines with null).

File: testinfra/modules/x509.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    @cached_property
+    def _openssl(self):
+        return self.find_command('openssl')
+
+    @property
+    def issuer(self):
+        stdout = self.check_output(
+            '%s x509 -in %s -noout -issuer', self._openssl, self.name
+        )
+        return stdout
```
Comment (philpep) [inline testinfra/modules/x509.py]: `issuer` and `subject` doesn't work for me:

```
import testinfra
host = testinfra.get_host('local://')
c = host.x509('/etc/ssl/certs/ssl-cert-snakeoil.pem')
for a in ('issuer', 'email', 'subject', 'enddate', 'serial'):
    print(a, repr(getattr(c, a)))
```

result in

```
issuer 'issuer=CN = aldo.in.philpep.org'
email ''
subject 'subject=CN = aldo.in.philpep.org'
enddate datetime.datetime(2023, 10, 31, 22, 51, 18)
serial 'A18E6C82C49B31EA'
```

I'm using openssl 1.1.1c-1 on debian buster.

**Comment by varac** [comment]
@julcollas Ping again, see above review from @philpep.

File: testinfra/modules/environment.py (hunk 7 above)
Comment (julcollas) [reply testinfra/modules/environment.py]: added in #491

**Comment by julcollas** [comment]
I created a new #491 only for environment module.

**Comment by julcollas** [comment]
I created a new #491 only for x509 module.
I close the PR

**Pull request closed without merging**